src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseSign1Message.cs

```diff
@@ -0,0 +1,283 @@
+
+        [UnsupportedOSPlatform("browser")]
+        public bool Verify(RSA key, ReadOnlySpan<byte> content) =>  VerifyRSA(key, _content == null ? content : throw new CryptographicException(SR.Sign1VerifyContentWasEmbedded));
+
+        [UnsupportedOSPlatform("browser")]
+        private bool VerifyECDsa(ECDsa key, ReadOnlySpan<byte> content)
+        {
+            PrepareForVerify(content, out int alg, out byte[] toBeSigned);
+            HashAlgorithmName hashAlgorithm = GetHashAlgorithmFromCoseAlgorithmAndKeyType(alg, KeyType.ECDsa);
+            return key.VerifyData(toBeSigned, _signature, hashAlgorithm);
+        }
+
+        [UnsupportedOSPlatform("browser")]
+        private bool VerifyRSA(RSA key, ReadOnlySpan<byte> content)
+        {
+            PrepareForVerify(content, out int alg, out byte[] toBeSigned);
+            HashAlgorithmName hashAlgorithm = GetHashAlgorithmFromCoseAlgorithmAndKeyType(alg, KeyType.RSA);
+            return key.VerifyData(toBeSigned, _signature, hashAlgorithm, RSASignaturePadding.Pss);
+        }
+
+        private void PrepareForVerify(ReadOnlySpan<byte> content, out int alg, out byte[] toBeSigned)
+        {
+            ReadOnlyMemory<byte> encodedAlg = GetCoseAlgorithmFromBuckets(ProtectedHeader, UnprotectedHeader);
+            alg = DecodeCoseAlgorithmHeader(encodedAlg) ?? throw new CryptographicException(SR.Sign1VerifyAlgHeaderWasIncorrect);
+            toBeSigned = CreateToBeSigned(SigStructureCoxtextSign1, _protectedHeaderAsBstr, content);
+        }
+
+        private static ReadOnlyMemory<byte> GetCoseAlgorithmFromBuckets(CoseHeaderMap protectedHeaders, CoseHeaderMap unprotectedHeaders)
+        {
+            // https://datatracker.ietf.org/doc/html/rfc8152#section-3.1 alg:
+            // This parameter MUST be authenticated where the ability to do so exists.
+            // This authentication can be done either by placing the header in the protected header bucket or as part of the externally supplied data.
+            // Example of an Algorithm header placed in the unprotected bucket https://github.com/cose-wg/Examples/blob/master/sign1-tests/sign-pass-01.json
+            CoseHeaderLabel label = CoseHeaderLabel.Algorithm;
+            if (!protectedHeaders.TryGetEncodedValue(label, out ReadOnlyMemory<byte> encodedAlg))
+            {
+                if (!unprotectedHeaders.TryGetEncodedValue(label, out encodedAlg))
+                {
+                    throw new CryptographicException(SR.Sign1SignAlgIsRequired);
+                }
+            }
+
+            return encodedAlg;
+        }
+
+        private static int? DecodeCoseAlgorithmHeader(ReadOnlyMemory<byte> encodedAlg)
+        {
+            var reader = new CborReader(encodedAlg);
+            CborReaderState state = reader.PeekState();
+
+            if (state == CborReaderState.NegativeInteger || state == CborReaderState.UnsignedInteger)
+            {
+                int alg = reader.ReadInt32();
```

Comment:
If this is a value that doesn't fit in Int32 then we'll throw a COSE exception on this line, vs the CryptographicException we'd throw for it being the int.MaxValue.

Reply:
On `CoseMessage.DecodeSign1`, if CborReader throws we catch the CborContentException and we throw a CryptographicException with the former being the inner exception. Perhaps we should wrap similarly with try-catch the places where we read CBOR in this and other methods.

Reply:
FWIW, this was actually throwing OverflowException.